assets/javascripts/discourse/components/chat-composer-uploads.js

```diff
@@ -0,0 +1,141 @@
+import Component from "@ember/component";
+import { clipboardHelpers } from "discourse/lib/utilities";
+import { run } from "@ember/runloop";
+import { action } from "@ember/object";
+import { inject as service } from "@ember/service";
+import UppyMediaOptimization from "discourse/lib/uppy-media-optimization-plugin";
+import discourseComputed, { bind } from "discourse-common/utils/decorators";
+import UppyUploadMixin from "discourse/mixins/uppy-upload";
+
+export default Component.extend(UppyUploadMixin, {
+  classNames: ["chat-composer-uploads"],
+  mediaOptimizationWorker: service(),
+  id: "chat-composer-uploader",
+  type: "chat-composer",
+  uploads: null,
+  useMultipartUploadsIfAvailable: true,
+
+  init() {
+    this._super(...arguments);
+    this.setProperties({
+      uploads: [],
+      fileInputSelector: `#${this.fileUploadElementId}`,
+    });
+    this.appEvents.on("chat-composer:load-uploads", this, "_loadUploads");
+  },
+
+  didInsertElement() {
+    this._super(...arguments);
+    this.composerInputEl = document.querySelector(".chat-composer-input");
+    this.composerInputEl?.addEventListener("paste", this._pasteEventListener);
+  },
+
+  willDestroyElement() {
+    this._super(...arguments);
+    this.appEvents.off("chat-composer:load-uploads", this, "_loadUploads");
+    this.composerInputEl?.removeEventListener(
+      "paste",
+      this._pasteEventListener
+    );
+  },
+
+  uploadDone(upload) {
+    this.uploads.pushObject(upload);
+    this.onUploadChanged(this.uploads);
+  },
+
+  @discourseComputed("uploads.[]", "inProgressUploads.[]")
```

Comment:
wonder if you could direct track .length here?

Reply:
Yeah good point :+1: